Feature/freesurfer v6 latest

tbbrown opened the pull request:
Prior to issuing this pull request, the <code>master</code> branch was merged into the <code>feature/freesurfer_v6_latest</code> branch both to make tests run using the <code>feature/freesurfer_v6_latest</code> branch use the latest code in <code>master</code> along with the changes in <code>feature/freesurfer_v6_latest</code> and to (hopefully) make pulling the changes from <code>feature/freesurfer_v6_latest</code> into the <code>master</code> easier.

The following is a general description of the significant changes to be found in the <code>feature/freesurfer_v6_latest</code> branch that is to be merged into the <code>master</code> with this pull request.

**Structural Preprocessing**

1. <code>FreeSurferPipeline.sh</code> has been re-written to use FreeSurfer v6.0.0 and Doug Greve's custom scripts
2. Doug Greve's custom scripts are in the <code>FreeSurfer/custom</code> directory
3. <code>FreeSurferPipeline-v5.3.0-HCP.sh</code> is kept around in case someone needs/wants to use it
4. The custom FreeSurfer scripts are now found on the <code>PATH</code>, full path specifications for these tools are not used.
    - The function <code>configure_custom_tools</code> in <code>FreeSurferPipeline.sh</code> checks to see if all 3 of the custom scripts are already findable (using the <code>which</code> command). If all three can be located, then it is assumed that the user has externally configured the <code>PATH</code> so that the custom scripts can be found. So <code>PATH</code> is not then altered.
    - If any one of the custom scripts is not findable, then <code>PATH</code> is modified so that <code>${HCPPIPEDIR}/FreeSurfer/custom</code> is at the beginning of the <code>PATH</code> making the custom scripts findable.
    - If <code>PATH</code> is modified, warning messages are output to let the user know that <code>PATH</code> is being changed. 
5. <code>FreeSurferPipeline.sh</code> checks that a valid version of FreeSurfer is being used
    - The <code>validate_freesurfer_version</code> function checks to make sure that FreeSurfer version 6.0.0 or later is being used
    - The <code>validate_freesurfer_version</code> function "parses" the contents of the <code>${FREESURFER_HOME}/build-stamp.txt</code> file to determine the FreeSurfer version in use
    - If the first number in the X.Y.Z version number found in that file is less than 6, then the script aborts with a message indicating that FreeSurfer 6.0.0 or greater is required 

**Functional Preprocessing**

1. Matt's latest changes to the generation of Spin Echo Bias field outputs in fMRIVolume preprocessing have been incorporated into <code>GenericfMRIVolumeProcessingPipeline.sh</code>

**Diffusion Preprocessing**

1. The mechanism for locating the eddy binary to use in the <code>DiffusionPreprocessing/scripts/run_eddy.sh</code> script has been reworked. These changes are included in the <code>determine_eddy_tools_to_use</code> and <code>determine_eddy_tools_for_supported_six_series</code> functions in the script.

    The general idea is:

    - FSL 5.0.7 or greater is required
    - FSL 5.0.7 and 5.0.8
        - non-GPU version of eddy is <code>${FSLDIR}/bin/eddy</code>
        - GPU version is <code>${FSLDIR}/bin/eddy.gpu</code>
    - FSL 5.0.9, 5.0.10, and 5.0.11 
        - non-GPU version: <code>${FSLDIR}/bin/eddy_openmp</code>
        - GPU version: <code>${FSLDIR}/bin/eddy_cuda</code>
    - FSL 6.0.0 is unsupported - error message is generated and processing aborted
    - FSL 6.0.1 and newer
        - non-GPU version: <code>${FSLDIR}/bin/eddy_openmp</code>
        - GPU-version (in the following order of precedence)
            - <code>${FSLDIR}/bin/eddy_cudaX.Y</code> where <code>X.Y</code> is specified as the <code>--cuda-version=</code> parameter value if given
            - <code>${FSLDIR}/bin/eddy</code> if it is found (recommended to be a symbolic link to the right <code>eddy_cudaX.Y</code> file)
            - <code>${FSLDIR}/bin/eddy_cuda</code> if it is found (recommended to be a symbolic link to the right <code>eddy_cudaX.Y</code> file

    The use of the GPU enabled version of eddy is the default for <code>DiffPreprocPipeline_Eddy.sh</code> and the <code>--cuda-version=</code> paramter can be specified to that script to be passed on down the call chain to the <code>run_eddy.sh</code> script.

**MR+FIX**

1. <code>hcp_fix_multi_run</code> checks that FSL 6.0.1 or greater is being used and aborts with an error message if that is not the case.

2. <code>ReApplyFixMultiRunPipeline.sh</code> also checks that FSL 6.0.1 or greater is being used.

mharms commented:
I'm not seeing my latest commit https://github.com/mharms/HCPpipelines/commit/23bbcf0e95454b0489cfe5370a7a13e29383b623 regarding the -conf2hires flag in FreeSurfer.sh. I had issued a PR, and I thought Matt approved it, so I'm not sure why it isn't in here.

mharms commented on FreeSurfer/custom/README:
```diff
@@ -0,0 +1,7 @@
+${HCPPIPEDIR}/FreeSurfer/custom, is on your PATH before running
+FreeSurferPipeline.sh. Run recon-all.v6.hires as you normally would
+WITHOUT -hires or -cm but adding -conf2hires.
+
```
This README might be a little confusing as currently written:
1) While the particular scripts in custom do not require a T2 scan, the FreeSurferPipeline.sh script itself does.
2) Perhaps delete the last sentence -- users aren't running recon-all.v6.hires directly themselves. Rather it is invoked through FreeSurferPipeline.sh (with the appropriate flags set therein).

tbbrown replied:
Looks like you've fixed that README file.

mharms commented:
I updated my branch, and addressed the issues that I mentioned above in my standing PR #95.  (I guess @glasserm reviewed it in its earlier state, but it didn't get merged in).

mharms commented:
@tbbrown Thanks for the very helpful summary at the top of the PR.  Overall, I think we are in pretty good shape.  Other than #95 needing to get merged, here are the loose ends that I'm aware of:

1. I'm waiting for the Diffusion test run to complete, but I suspect that we'll need to include this specific commit https://github.com/MichielCottaar/Pipelines/commit/fc3e20e409b0f34709e6374f7b7077f09fc50833#diff-3321c68683f48df3fcd1ae19774d1bd5.  As for the broader #73 from @MichielCottaar, I think that should wait, as we need to do some additional testing/review.  Plus, we are going to need other additional modifications to `DiffusionPreprocessing` anyway to support use of the new motion-by-susceptibility correction in 'eddy'.  So, in principle, we can lump all the `DiffusionPreprocessing` changes/additions together.
2. The issue of supported versions of tools is starting to get a little messy, mostly because FSL 6.0.0 really shouldn't be used for anything.  Do we want to switch to the assumption that FSL 6.0.1 is the "new" minimum version (and update all the comments in all the scripts accordingly)?  That wouldn't necessarily require that we explicitly _abort_ all scripts if the FSL version isn't 6.0.1 or later, but would help to reset expectations about what version combinations we can reasonably hope to support.  And we probably need to do something with `fMRIVolume` similar to what @tbbrown did with `DiffusionPreprocessing`, since `fMRIVolume` really should error out if someone is trying to use FSL 6.0.0.
3. Compiled Matlab support for MR-FIX is not going to make it in this pipeline release.  Hopefully we can get that added as an enhancement to the master branch in the near future though.

tbbrown commented:
> I'm not seeing my latest commit [mharms@23bbcf0](https://github.com/mharms/HCPpipelines/commit/23bbcf0e95454b0489cfe5370a7a13e29383b623) regarding the -conf2hires flag in FreeSurfer.sh. I had issued a PR, and I thought Matt approved it, so I'm not sure why it isn't in here.

From looking at the PR, I see that Matt approved it, but it didn't get merged. I too thought that the approval meant that it had been merged. I merged PR #95 just a few moments ago. That pulled your changes into the feature/freesurfer_v6_latest branch so that this PR #96 should get those changes pulled into the master branch.

tbbrown commented:
Since it already had Matt's approval, I went ahead and merged PR #95. 
That brought those changes into the feature/freesurfer_v6_latest branch 
so they will be merged into the master with this pull request (#96).

@mharms: I started to address the other comments you made (README 
updates and other changes). But it appears that you have already address 
them. Have you addressed all the comments you made or are there some 
that I still need to look at?

Re #1:

    The Diffusion test run has completed (see separate email for where
    the results are located)

    I'm unsure if you are saying that we really need to get the specific
    MichielCottaar commit you listed into the repository before we do a
    v4.0.0 release or whether you are saying we can wait to do that
    before a future release that addresses the motion-by-susceptibilty
    correction issue.

   Tim

On 3/7/19 14:45, Michael Harms wrote:

> @tbbrown <https://github.com/tbbrown> Thanks for the very helpful 
> summary at the top of the PR. Overall, I think we are in pretty good 
> shape. Other than #95 
> <https://github.com/Washington-University/HCPpipelines/pull/95> 
> needing to get merged, here are the loose ends that I'm aware of:
>
>  1. I'm waiting for the Diffusion test run to complete, but I suspect
>     that we'll need to include this specific commit
>     MichielCottaar@fc3e20e#diff-3321c68683f48df3fcd1ae19774d1bd5
>     <https://github.com/MichielCottaar/Pipelines/commit/fc3e20e409b0f34709e6374f7b7077f09fc50833#diff-3321c68683f48df3fcd1ae19774d1bd5>.
>     As for the broader #73
>     <https://github.com/Washington-University/HCPpipelines/pull/73>
>     from @MichielCottaar <https://github.com/MichielCottaar>, I think
>     that should wait, as we need to do some additional testing/review.
>     Plus, we are going to need other additional modifications to
>     |DiffusionPreprocessing| anyway to support use of the new
>     motion-by-susceptibility correction in 'eddy'. So, in principle,
>     we can lump all the |DiffusionPreprocessing| changes/additions
>     together.
>  2. The issue of supported versions of tools is starting to get a
>     little messy, mostly because FSL 6.0.0 really shouldn't be used
>     for anything. Do we want to switch to the assumption that FSL
>     6.0.1 is the "new" minimum version (and update all the comments in
>     all the scripts accordingly)? That wouldn't necessarily require
>     that we explicitly /abort/ all scripts if the FSL version isn't
>     6.0.1 or later, but would help to reset expectations about what
>     version combinations we can reasonably hope to support. And we
>     probably need to do something with |fMRIVolume| similar to what
>     @tbbrown <https://github.com/tbbrown> did with
>     |DiffusionPreprocessing|, since |fMRIVolume| really should error
>     out if someone is trying to use FSL 6.0.0.
>  3. Compiled Matlab support for MR-FIX is not going to make it in this
>     pipeline release. Hopefully we can get that added as an
>     enhancement to the master branch in the near future though.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/Washington-University/HCPpipelines/pull/96#issuecomment-470687341>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AFt4x3S91hakGEdaL0I_ZQJ1-56N7xcWks5vUXqDgaJpZM4bjrBZ>.
>
-- 
/Timothy B. Brown
Business & Technology Application Analyst III
Pipeline Developer (Connectome Coordination Facility)
tbbrown(at)wustl.edu
/
------------------------------------------------------------------------
The material in this message is private and may contain Protected 
Healthcare Information (PHI). If you are not the intended recipient, be 
advised that any unauthorized use, disclosure, copying or the taking of 
any action in reliance on the contents of this information is strictly 
prohibited. If you have received this email in error, please immediately 
notify the sender via telephone or return mail.

mharms commented:
Re (1) from https://github.com/Washington-University/HCPpipelines/pull/96#issuecomment-470687341, that specific commit fc3e20e409 from @MichielCottaar needs to go into the repo now.  I've confirmed that the problem mentioned in that commit is present in our test `DiffusionPreprocessing` run.  I also sent an email to Oxford, inquiring why they changed the behavior of `bet`.  We also need to make sure that none of the other uses of `bet` in _any_ pipeline script are operating on a 4D file.

mharms commented:
I used GitHub's search feature to search for all instances of "bet" within the repo.  The only instance where `bet` is getting applied to a 4D file is in `eddy_postproc.sh`, so that's the only instance we need to fix.

mharms commented:
@coalsont  One other thing I just remembered is that we don't have any guidance about how users are actually supposed to run MR-FIX in the high-memory-requirement situation.  i.e., we don't have a container that they can use currently (unless you plan to rebuild yours using Octave and publish it somewhere); we don't have any notes in ICAFIX/README.md about building your own high memory version of Octave; and we don't (yet) have the compiled matlab ready.

Unless we resolve this, we will be "announcing" MR-FIX as a new feature, but users won't actually be able to run it. :(

coalsont commented:
MR FIX will run in interpreted matlab, it is just that that isn't particularly cluster-friendly.

mharms commented:
@coalsont I think that any guidance we can add to ICAFIX/README regarding how one would build a high memory version of Octave would be a helpful inclusion.

coalsont commented:
I have added the octave build instructions to the ICAFIX README in master.

glasserm commented:
Thanks Tim
